File: Algorithms/141. Linked List Cycle/golang/solution.hash-table.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycle(head *ListNode) bool {
	var nodesSeen = make(map[*ListNode]bool)
	for head != nil {
		if nodesSeen[head] {
			return true
		}
		nodesSeen[head] = true
		head = head.Next
	}
	return false
}

File: Algorithms/141. Linked List Cycle/kotlin/solution.hash-table.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun hasCycle(head: ListNode?): Boolean {
        var nodesSeen = mutableSetOf<ListNode>()
        var node = head
        while (node != null) {
            if (nodesSeen.contains(node)) {
                return true
            }
            nodesSeen.add(node)
            node = node.next
        }
        return false
    }
}